### ISSUE TITLE
Importing mteb is slow

### ISSUE THREAD

**Issue by Muennighoff** [embeddings-benchmark/mteb#1463]
mteb takes 30s while transformers takes 6s (https://colab.research.google.com/drive/166j9HRQF8hAqwiiYpMDvYoznm1v-veOC?usp=sharing). Probably because of the many `*` imports 🤔

**Comment by KennethEnevoldsen** [comment]
Yeah we can def. reduce the number of `*` imports. I could also imagine that it might be the many TaskMetadata objects that take some time as well - we should probably profile it to see where the problem is.

**Comment by KennethEnevoldsen** [comment]
Yes v2.0.0 is the best place to start

**Comment by Napuh** [comment]
At first glance the star imports are not taking that much time to load. Using `py-spy` I can see that `datasets` and `sentence_transformers` are taking the mayority of the time spent importing:

![profile](https://github.com/user-attachments/assets/27d3ff98-f1f0-4c8c-b325-fbd25f3f68c1)

I can get rid of some star imports, but this will not make any effect on import speed :(

**Comment by Muennighoff** [comment]
Hm as mentioned here https://github.com/embeddings-benchmark/mteb/issues/102 we can make sentence-transformers (and also torch) an optional dependency?

**Comment by Napuh** [comment]
Also, on `v2.0.0` i hace not experienced the 30s import time. For me it just took 5-ish seconds.

**Comment by KennethEnevoldsen** [comment]
For me it is still slow on v2.0.0 - So it might depend on versions or python instance -- @Napuh are you using python 3.9?

An optional thing we could do is to move the datasets import into the load_dataset function, but I think a PR adding a new dataset will likely not do this so I don't think it is a viable solution.

We could make sentence transformers an optional dependency but the error message would have to be good to avoid confusion. v2.0.0 is probably the version to do it in

**Comment by Samoed** [comment]
When working on https://github.com/embeddings-benchmark/mteb/pull/1362, I encountered weird issues with the `logging` module due to circular imports. I believe removing `*` imports would resolve the problem. @Napuh

**Comment by ayush1298** [comment]
@Samoed , should we mark this as complete?

**Comment by Samoed** [comment]
It still kind slow, so I don't know. @KennethEnevoldsen What do you think?

**Comment by KennethEnevoldsen** [comment]
So def. possible to speed it 

Adding this [rule](https://docs.astral.sh/ruff/rules/#flake8-type-checking-tc) could likely help removing a lot of unec. import during runtime.

Also tested out if simply converting import speeds it up and it does. 

I.e.
```py
import datasets
# to
from datasets import load_datasets # faster
```

So that would also be an option (thus preventing full large library imports like torch, datasets and sentence_transformers.

**Comment by Samoed** [comment]
I think we also have a lot of metadata validation and because of this importing can be slow

**Comment by KennethEnevoldsen** [comment]
Yeah, I was actually thinking of fixing it in models such that creating the model registry happens later.

There are a couple of approaches:

Current:
```py
import mteb # building registry happens as import, this includes validation

mteb.get_model_meta(name)
```

Approach 2: Move validation to tests
```py
import mteb # building registry happens as import, but no validation
mteb.get_model_meta(name)

# during meta definition
meta = ModelMeta.model_construct(**kwargs) # This looks horrible we might be able to do it in another way

# During testing
meta.validate_model(meta)
```

Approach 3: Move import to get_model_meta
```py
import mteb # models registry isn't built, and the model's folder isn't imported.

mteb.get_model_meta(name) # will built registry if not built otherwise fetch from it
```
The problems is that might lead to a validation error only during `mteb.get_model_meta`.

Personally, I am a fan of 2. if we can figure out a way to do it so that the definition looks nice. However, that might require us to move away from pydantic.

**Comment by Samoed** [comment]
This also should be done for tasks too

**Comment by KennethEnevoldsen** [comment]
I created two sub-issues here. I would, however, note that when profiling, I see that the majority of the time is spent on imports of other libraries, not on validation.

**Comment by KennethEnevoldsen** [comment]
So seems like we have attempted both solutions here without noticeable gains. 

Generally, it seems like the majority of the time is from imports:
- sentence_transformers import - especially CrossEncoder
- transformers library imports
- torch imports

A solution could be to make sentence_transformers an optional import/install. We could probably also make transformers an optional install/import.

There are a few minor functions that take some time e.g., when building the language grid for Flores (though these are so small that fixing all of them wouldn't even get us close to actually making a package optional)

These can be handled after v2, so I will just remove the v2 tag.

**Comment by Muennighoff** [comment]
Nice analysis! +1 to making dependencies optional; we can even make torch optional I think (would be interesting to know how many people use mteb to only benchmark API endpoints where you wouldn't need any of these)

**Comment by KennethEnevoldsen** [comment]
Yeah, we could make some of these optional.

A quick overview:
- For SentenceTransformers, it should be fairly easy to do. Though it would especially require restructuring the metadata computation for new models
- For torch, it is likely impossible as we use the dataloader in v2 to support efficient loading of other modalities. It is also required for exception API computation (though do see #941)
- For transformers, the main stopper is that we use it to set the seed in tasks (we could reimplement that functionality). In addition, multiple models use transformers
